server/plugin.go

```diff
@@ -97,281 +85,85 @@ func (p *Plugin) OnActivate() error {
 			"user:read",
 			"meeting:write",
 			"webinar:write",
 			"recording:write"},
-	}, nil
-}
-
-type ZoomUserInfo struct {
-	ZoomEmail string
-
-	// Zoom OAuth Token, ttl 15 years
-	OAuthToken *oauth2.Token
-
-	// Mattermost userID
-	UserID string
-
-	// Zoom userID
-	ZoomID string
-}
-
-type AuthError struct {
-	Message string `json:"message"`
-	Err     error  `json:"err"`
-}
-
-func (ae *AuthError) Error() string {
-	errorString, _ := json.Marshal(ae)
-	return string(errorString)
-}
-
-func (p *Plugin) storeZoomUserInfo(info *ZoomUserInfo) error {
-	config := p.getConfiguration()
-
-	encryptedToken, err := encrypt([]byte(config.EncryptionKey), info.OAuthToken.AccessToken)
-	if err != nil {
-		return err
-	}
-
-	info.OAuthToken.AccessToken = encryptedToken
-
-	jsonInfo, err := json.Marshal(info)
-	if err != nil {
-		return err
-	}
-
-	if err := p.API.KVSet(zoomTokenKey+info.UserID, jsonInfo); err != nil {
-		return err
-	}
-
-	if err := p.API.KVSet(zoomTokenKeyByZoomID+info.ZoomID, jsonInfo); err != nil {
-		return err
 	}
-
-	return nil
 }
 
-func (p *Plugin) getZoomUserInfo(userID string) (*ZoomUserInfo, error) {
-	config := p.getConfiguration()
-
-	var userInfo ZoomUserInfo
-
-	infoBytes, appErr := p.API.KVGet(zoomTokenKey + userID)
-	if appErr != nil || infoBytes == nil {
-		return nil, errors.New("must connect user account to Zoom first")
-	}
-
-	err := json.Unmarshal(infoBytes, &userInfo)
+func (p *Plugin) authenticateAndFetchZoomUser(user *model.User) (*zoom.User, *zoom.AuthError) {
+	zoomClient, err := p.getActiveClient(user)
 	if err != nil {
-		return nil, errors.New("unable to parse token")
-	}
-
-	unencryptedToken, err := decrypt([]byte(config.EncryptionKey), userInfo.OAuthToken.AccessToken)
-	if err != nil {
-		log.Println(err.Error())
-		return nil, errors.New("unable to decrypt access token")
-	}
-
-	userInfo.OAuthToken.AccessToken = unencryptedToken
-
-	return &userInfo, nil
-}
-
-func (p *Plugin) authenticateAndFetchZoomUser(userID, userEmail, channelID string) (*zoom.User, *AuthError) {
-	var zoomUser *zoom.User
-	var clientErr *zoom.ClientError
-	var err error
-	config := p.getConfiguration()
-
-	// use OAuth
-	if config.EnableOAuth {
-		zoomUserInfo, apiErr := p.getZoomUserInfo(userID)
-		oauthMsg := fmt.Sprintf(
-			zoomOAuthMessage,
-			*p.API.GetConfig().ServiceSettings.SiteURL, channelID)
-
-		if apiErr != nil || zoomUserInfo == nil {
-			return nil, &AuthError{Message: oauthMsg, Err: apiErr}
-		}
-		zoomUser, err = p.getZoomUserWithToken(zoomUserInfo.OAuthToken)
-		if err != nil || zoomUser == nil {
-			return nil, &AuthError{Message: oauthMsg, Err: apiErr}
-		}
-	} else {
-		// use personal credentials
-		zoomUser, clientErr = p.zoomClient.GetUser(userEmail)
-		if clientErr != nil {
-			includeEmailInErr := fmt.Sprintf(zoomEmailMismatch, userEmail)
-			return nil, &AuthError{Message: includeEmailInErr, Err: clientErr}
+		return nil, &zoom.AuthError{
+			Message: fmt.Sprintf(zoom.OAuthPrompt, p.siteURL),
```

Comment:
Do you mind adding a comment here to explain that an error getting the client means we are using OAuth and the user is not connected?

Reply:
Or any kind of reordering that keeps the code clearer is also an alternative.